rascaline/src/calculators/neighbor_list.rs

```diff
@@ -0,0 +1,620 @@
+                        let second_grad_sample_i = gradient.samples.position(&[
+                            sample_i.into(), system_i.into(), pair.first.into()
+                        ]).expect("missing gradient sample");
+
+                        let array = gradient.data.as_array_mut();
+
+                        array[[first_grad_sample_i, 0, 0, 0]] = -1.0;
+                        array[[first_grad_sample_i, 1, 1, 0]] = -1.0;
+                        array[[first_grad_sample_i, 2, 2, 0]] = -1.0;
+
+                        array[[second_grad_sample_i, 0, 0, 0]] = 1.0;
+                        array[[second_grad_sample_i, 1, 1, 0]] = 1.0;
+                        array[[second_grad_sample_i, 2, 2, 0]] = 1.0;
+                    }
+                }
+            }
+        }
+
+        return Ok(());
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use approx::assert_relative_eq;
+    use equistore::{LabelValue, LabelsBuilder, Labels};
+
+    use crate::systems::test_utils::{test_systems, test_system};
+    use crate::Calculator;
+
+    use super::NeighborList;
+    use super::super::CalculatorBase;
+
+    #[test]
+    fn half_neighbor_list() {
+        let mut calculator = Calculator::from(Box::new(NeighborList{
+            cutoff: 2.0,
+            full_neighbor_list: false,
+        }) as Box<dyn CalculatorBase>);
+
+        let mut systems = test_systems(&["water"]);
+
+        let descriptor = calculator.compute(&mut systems, Default::default()).unwrap();
+
+        assert_eq!(*descriptor.keys(), Labels::new(
+            ["species_first_atom", "species_second_atom"],
+            &[[-42, 1], [1, 1]]
+        ));
+
+        // O-H block
+        let block = descriptor.blocks()[0].values();
+        assert_eq!(*block.properties, Labels::new(["distance"], &[[0]]));
+
+        assert_eq!(block.components.len(), 1);
+        assert_eq!(*block.components[0], Labels::new(["pair_direction"], &[[0], [1], [2]]));
+
+        assert_eq!(*block.samples, Labels::new(
+            ["structure", "pair_id", "first_atom", "second_atom"],
+            // we have two O-H pairs
+            &[[0, 0, 0, 1], [0, 1, 0, 2]]
+        ));
+
+        let array = block.data.as_array();
+        let expected = &ndarray::arr3(&[
+            [[0.0], [0.75545], [-0.58895]],
+            [[0.0], [-0.75545], [-0.58895]]
+        ]).into_dyn();
+        assert_relative_eq!(array, expected, max_relative=1e-6);
+
+        // H-H block
+        let block = descriptor.blocks()[1].values();
+        assert_eq!(*block.samples, Labels::new(
+            ["structure", "pair_id", "first_atom", "second_atom"],
+            // we have one H-H pair
+            &[[0, 2, 1, 2]]
+        ));
+
+        let array = block.data.as_array();
+        let expected = &ndarray::arr3(&[
+            [[0.0], [-1.5109], [0.0]]
+        ]).into_dyn();
+        assert_relative_eq!(array, expected, max_relative=1e-6);
+    }
+
+    #[test]
+    fn full_neighbor_list() {
```

Comment:
I don't really see here where you test the `full_neighbor_list`. The test looks exactly the same as the `half_neighbor_list`.

Reply:
the `full_neighbor_list` hyper is true, and there are more pairs in this test. but yes, overall the structure is very similar

Reply:
Okay, maybe I just oversaw the changes this morning.